lib/ts/ingredients/smsdelivery/types.ts

```diff
@@ -0,0 +1,30 @@
+/* Copyright (c) 2021, VRAI Labs and/or its affiliates. All rights reserved.
+ *
+ * This software is licensed under the Apache License, Version 2.0 (the
+ * "License") as published by the Apache Software Foundation.
+ *
+ * You may not use this file except in compliance with the License. You may
+ * obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+import OverrideableBuilder from "supertokens-js-override";
+
+export type SmsDeliveryInterface<T> = {
+    sendSms: (input: T & { userContext: any }) => Promise<void>;
+};
+
+/**
+ * config class parameter when parent Recipe create a new SmsDeliveryRecipe object via constructor
+ */
+export interface TypeInput<T> {
+    service: SmsDeliveryInterface<T>;
```

Comment:
Shouldn't `service` be optional too? (Same goes for email sending service as well). The reason why it should be optional here is that:
- The user may want to override fully the way to send SMS / email and then this service param makes no sense anymore.
- You are anyway handling the case when the server param is `undefined` in the `getSmsDeliveryConfig()` function.

Finally, the return type of `getSmsDeliveryConfig` should not be same as `TypeInput` since there, we want `service` to always be defined. So you will require two types:
```ts
export interface TypeInput<T> {
    service?: SmsDeliveryInterface<T>;
    override?: (
        originalImplementation: SmsDeliveryInterface<T>,
        builder: OverrideableBuilder<SmsDeliveryInterface<T>>
    ) => SmsDeliveryInterface<T>;
}
```
and
```ts
export interface TypeInputWithService<T> {
    service: SmsDeliveryInterface<T>;
    override?: (
        originalImplementation: SmsDeliveryInterface<T>,
        builder: OverrideableBuilder<SmsDeliveryInterface<T>>
    ) => SmsDeliveryInterface<T>;
}
```

The input to the `constructor` in `/ingredients/smsdelivery/index.ts` should be `TypeInputWithService`

Reply:
made the changes for smsDelivery. changes for emailDelivery would be done in other PR

Reply:
The changes for SMS delivery for this are fine. Need to make same for email delivery. Please do that in this PR only